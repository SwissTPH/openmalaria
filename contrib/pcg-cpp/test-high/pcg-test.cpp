/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
