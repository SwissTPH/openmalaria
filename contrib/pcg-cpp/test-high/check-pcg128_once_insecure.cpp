#define RNG pcg128_once_insecure
#define TWO_ARG_INIT 1

#include "pcg-test.cpp"

