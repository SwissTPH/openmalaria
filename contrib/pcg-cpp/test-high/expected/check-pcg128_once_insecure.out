pcg128_once_insecure:
      -  result:      128-bit unsigned int
      -  period:      2^128   (* 2^127 streams)
      -  size:        32 bytes

Round 1:
 128bit: 0x5f4ea96e8510af0686b1da1d72062b68 0x341b1cb1e675ec461304aa46c9853d39
	 0xcfdc46c17f1c9974a3670e9e0dd50358 0x02d273b87fe9110cf9090e529a7dae00
	 0x9b4e47fda576f0ddc85b9fd837996f2c 0x17cee59c8cb9c0a1606121f8e3919196
  Again: 0x5f4ea96e8510af0686b1da1d72062b68 0x341b1cb1e675ec461304aa46c9853d39
	 0xcfdc46c17f1c9974a3670e9e0dd50358 0x02d273b87fe9110cf9090e529a7dae00
	 0x9b4e47fda576f0ddc85b9fd837996f2c 0x17cee59c8cb9c0a1606121f8e3919196
  Coins: TTTHHHTTTHHHTTTTHHTTHHTHTHTTHHTHTTTTHHTTTHTHHTHTTTTHHTTTHHHTTTHTT
  Rolls: 4 4 1 5 3 1 1 3 4 5 4 6 2 5 2 5 3 1 3 3 1 4 6 2 5 4 4 3 3 6 6 3 2
   -->   rolling dice used 33 random numbers
  Cards: As 2s 8h Tc 3c 5h Ac 5c 9d 8d Td Ts 4c 7d 9c Jh 3d Th 8s 5s 2h 6d
	 7h 3h Qd 7c Qh Ks Js Kh 6s 3s Qc Ah 6h 9s Ad Kc 6c Kd 5d Jd 2d Jc
	 2c 4d Qs 9h 7s 8c 4h 4s

Round 2:
 128bit: 0x95a15dc187be75191773ba241e7a792a 0x7f9a9c8dc8ee5d30e41aed7117b0bc10
	 0x859f2aaeb7456bae36bac8d9432af525 0xd83833165d913fb0e0c78e2f3c850a38
	 0xa437ee7e51a835ebe3ad939c1c7ce70d 0x9a515b18be44d777a302fdced8c79e93
  Again: 0x95a15dc187be75191773ba241e7a792a 0x7f9a9c8dc8ee5d30e41aed7117b0bc10
	 0x859f2aaeb7456bae36bac8d9432af525 0xd83833165d913fb0e0c78e2f3c850a38
	 0xa437ee7e51a835ebe3ad939c1c7ce70d 0x9a515b18be44d777a302fdced8c79e93
  Coins: TTTTHTHTHHTHTHTTTTTHHTTHHHHTHTHHHHHHHTHHHTHHTHTTTHHHHTTHHTTTHTHTH
  Rolls: 6 5 3 3 2 3 3 2 5 2 6 6 6 4 2 2 5 5 2 2 4 1 5 3 1 5 5 6 2 4 5 2 1
   -->   rolling dice used 33 random numbers
  Cards: 3s Qd Qc 6s 7c 8d 5s 5c Kc 7h Kd Ad 4d 6c Kh Qh 2c Jc Th Jh 4c Ks
	 Td 8s Jd 2h Ac 9c 7s 8c 3c Ah 9h 3h 7d 8h Js Qs As 4s 6d 5d 9s 9d
	 5h 3d 6h Tc 4h Ts 2d 2s

Round 3:
 128bit: 0x6bfdb8dec74ac18ac96006593aed3b62 0xeef8b9c71cb3b932f04d5afa3f197bf1
	 0x904d15f0ead523cfce6f729cc913a50f 0x34d0abbd7c06a00798b5fc4fbb1e4aea
	 0xa7d1cc41ac8cf8d1802dce1b410fc8c3 0x1622c80e085b475ae3bac0a14f6e5033
  Again: 0x6bfdb8dec74ac18ac96006593aed3b62 0xeef8b9c71cb3b932f04d5afa3f197bf1
	 0x904d15f0ead523cfce6f729cc913a50f 0x34d0abbd7c06a00798b5fc4fbb1e4aea
	 0xa7d1cc41ac8cf8d1802dce1b410fc8c3 0x1622c80e085b475ae3bac0a14f6e5033
  Coins: HTTHTHTTTTTHTTTHHTHTHHTHHHHHHHHHTTTHTHTHTHHTTTTTTHHHHTHTTTTHHHHHH
  Rolls: 1 4 4 1 5 1 2 1 4 5 6 3 3 1 6 5 6 5 4 6 6 5 2 6 6 6 1 4 4 6 5 4 1
   -->   rolling dice used 33 random numbers
  Cards: Jh Th 3s 4d 7c Kd Kc 4h 8h 3c 9d 3h Ac 5h 9s 8d 6d 7h 5s Qs 5d Ah
	 Td 8s Ad 6s 7s 7d Ts 2h 5c As Tc 3d Js 4s 2s Jc 2d Ks 6h 2c 4c 6c
	 Qc Qd 9c Kh 8c Jd Qh 9h

Round 4:
 128bit: 0x776dd6f62f789d2368da679de81de48a 0x325b1acf6c0987ad7ee3c031fa0aa440
	 0x4c2fa58857bd28a36eb1663983530403 0x2da890444508568bfec4d7a9a7aec823
	 0x2d0d7462ff348734bce221c255ee9467 0xf27f193f4fdbda86460a42a962b8a2f9
  Again: 0x776dd6f62f789d2368da679de81de48a 0x325b1acf6c0987ad7ee3c031fa0aa440
	 0x4c2fa58857bd28a36eb1663983530403 0x2da890444508568bfec4d7a9a7aec823
	 0x2d0d7462ff348734bce221c255ee9467 0xf27f193f4fdbda86460a42a962b8a2f9
  Coins: HHHTTTTHHHHHTTTTTTTHHHTHHHHTTHTTTHTTTTHTHHHHTHHTTTHHHTHHTTHHHTHTH
  Rolls: 5 1 4 1 2 6 5 4 1 2 1 3 2 2 5 3 5 1 4 2 2 4 1 1 1 2 4 4 5 3 6 5 6
   -->   rolling dice used 33 random numbers
  Cards: 8h Ks 2s 5h Kh 2c Kc 9d 8c 9c 3s Jd Ts 5d 7c 8s Ah Jc 5s 4h Ad 4s
	 7d Qc 4c Jh 3d 6s Qh 7s 4d Tc 2h Ac 6c Js 3c Qd As 8d 9s 5c 9h 3h
	 6h Th Kd 6d 2d Td 7h Qs

Round 5:
 128bit: 0x2a51386ff58abc4f9e0d084cff42fe2f 0x4f5b3035738c6f7a63cd8347aae338ea
	 0xacdba2191e7281ad112aae00540d3fa1 0xa46125301827966c53968bc829afd6ec
	 0xc11d0f0a93287f5c1b9900eb6c5b6d90 0xd723aefa2d73f7b7e89ed17ea33cb420
  Again: 0x2a51386ff58abc4f9e0d084cff42fe2f 0x4f5b3035738c6f7a63cd8347aae338ea
	 0xacdba2191e7281ad112aae00540d3fa1 0xa46125301827966c53968bc829afd6ec
	 0xc11d0f0a93287f5c1b9900eb6c5b6d90 0xd723aefa2d73f7b7e89ed17ea33cb420
  Coins: HTTTTTHTHTHHHTHTTTHTHHTHHTHTTTHHTTHHHTTTTHTTHHTHHTHHHTTHHTHTHHHHH
  Rolls: 6 4 5 5 1 2 1 3 1 5 4 4 5 4 6 2 2 6 6 4 2 3 4 5 3 1 2 1 5 6 1 3 2
   -->   rolling dice used 33 random numbers
  Cards: 8s 8h Td 8c Js Jd 5s Kd 8d 9s 4d Ah 2d 4h As Th 4s 6s Qc Ad 9h 3h
	 Qh 5d 3s 9c 7c 3c Ts Kc 6c Jh 6d 2h 7d 9d Ac 5c Qs 2c 6h Ks 4c 3d
	 Kh 2s 7s Qd Jc 5h 7h Tc

